QModManager/Messages.cs

```diff
@@ -0,0 +1,71 @@
+﻿using System;
+using System.Linq;
+using System.Reflection;
+
+namespace QModManager
+{
+    public static partial class QModAPI
+    {
+        [AttributeUsage(AttributeTargets.Method, AllowMultiple = false, Inherited = false)]
+        public class MessageReceiverAttribute : Attribute
+        {
+            public QMod Sender;
+
+            public MessageReceiverAttribute(QMod sender)
+            {
+                Sender = sender ?? throw new ArgumentNullException("The provided mod is null!", nameof(sender));
+            }
+            public MessageReceiverAttribute(Assembly senderAssembly)
+            {
+                Sender = GetMod(senderAssembly, true, true) ?? throw new ArgumentException("The provided assembly is not a mod assembly!", nameof(senderAssembly));
+            }
+            public MessageReceiverAttribute(string senderID)
+            {
+                // Should we not want to throw an exception in case the mod is not present?
+                Sender = GetMod(senderID, true, true) ?? throw new ArgumentException($"No mod matching the provided ID \"{senderID}\" was found!", nameof(senderID));
+            }
+        }
+
+        [AttributeUsage(AttributeTargets.Method, AllowMultiple = false, Inherited = false)]
+        public class GlobalMessageRecieverAttribute : Attribute
+        {
+            public GlobalMessageRecieverAttribute() { }
+        }
+
+        public static void SendMessage(QMod mod, string message, params object[] data)
+        {
+            if (mod == null) return;
+            if (data == null) data = new object[] { };
+
+            QMod caller = GetMod(Assembly.GetCallingAssembly(), true);
+
+            mod.LoadedAssembly.GetTypes()
+                .SelectMany(t => t.GetMethods())
+                .Where(m => m.IsStatic)
+                .Where(m => m.GetParameters().Length == 3)
+                .Where(m => m.GetCustomAttributes(typeof(MessageReceiverAttribute), false)
+                    .Where(a => ((MessageReceiverAttribute)a).Sender == caller)
+                    .ToArray().Length > 0)
+                .ForEach(m => m.Invoke(null, new object[] { caller, message, data }));
```

Comment:
Can these methods not be detected and cached during the loading process?  
Having to find them all over again each time seems like an unnecessary performance hit.

If we already have a collection of all mod assemblies, why not expand the `QMod` class with cached references to the methods this needs to invoke?